[CI/Build]: Build container image when a new release is published

hickeyma opened the pull request:
Adds job to publish workflow to build container image when a new release is published. It also sets the `latest` tag to the image.

At the moment, this is being done manually. It will be more consistent and efficient to add it as a job to the publish workflow after package is published to PyPi.

Summary of the container image ([LMCache integrated with vLLM](https://hub.docker.com/r/lmcache/vllm-openai)) build workflow:

- Scheduled nightly build which pushes two tags 1. nightly-{date} and 2. `latest-nightly`
- Automated build when release is cut with two tags 1. Release tag (e.g. `v0.x.y`) and 2. `latest`

~TODO: Update the documentation on container image.~ Updated.

Fixes #686 
Partial: #594

~**Note to reviewers:** Update `publish.yml` to only run image publish when release is cut, when happy with workflow and ready for merging.~ Pushed [commit](https://github.com/LMCache/LMCache/pull/784/commits/a5467f8302c37d7370ab626d179fc923acf782d7) to remove the events for testing. Ready to merge.

hickeyma committed:
Add job to publish workflow to build container image

Build container image when a new release is published.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
@maobaolong @wwl2755 

Workflow running at the moment depending on `build-artifacts` job so that it runs each time I push a commit to PR (testing purposes only). I will change when ready for review to be dependent on `publish-pypi` job and only run when a release is cut.

The `publish-image` job ran [here](https://github.com/LMCache/LMCache/actions/runs/15537346963/job/43741013618?pr=784) showing how it built and pushed the image to DockerHub as shown below:


![Screenshot 2025-06-09 at 16 45 44](https://github.com/user-attachments/assets/6fddb4e3-58a8-4698-9236-4365166e1286)

At the moment I am using the following tags for the images:

- Tag of release just cut in the `publish-pypi` job. In this example I am using the current `v0.3.0` release tag.
- `latest` tag to point to the image (this would fix the issue raised in #686)

@ApostaC @YaoJiayi @sammshen This PR would automate the stable build of the container image and make sure the latest tag is not stale either. I know previously that date was used for the tag but wondering if it is better now to use the LMCache release tag to differentiate with nightly builds?

I would appreciate all your feedback. FYI, I am also going to update the documentation in this PR.

sammshen commented:
I believe all the dated tags were manually built previously independent of release. 

@hickeyma the tagging design of having one for the release (like you did for v0.3.0) and also pointing "latest" toward that on releases feels intuitive and is a great bug fix!!  

To summarize the docker workflows:
1. already pushing two tags 1. nightly-{date} and 2. latest-nightly every day
2. now automating pushing two tags 1. v0.x.x and 2. latest on release

is this a correct understanding?

hickeyma commented:
Yes, that's correct @sammshen

hickeyma committed:
Update documentation for new container image generation

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

sammshen commented:
Looks great! Thanks for updating the documentation as well!

Siddhant-Ray reviewed (approved):
lgtm

hickeyma committed:
Remove GH events used for testing

Events set to test workflow on every commit. Refactored to perform
image build on release only as per requirement.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
FYI, pushed last [commit](https://github.com/LMCache/LMCache/pull/784/commits/a5467f8302c37d7370ab626d179fc923acf782d7) to remove the events for testing and configure for release only as per specification to run the image build.

hickeyma committed:
Fix the GH action linter error

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>
